Genre-recommendation example update (#14)
neo4j-product-examples/snowflake-graph-analytics: genre-recommendation/README.md

# Genre Recommendation with IMDB Data

## Overview

> Duration: 2

Reviewer: what does Duration: 2 mean?

Author: 2 min, it's aligned with https://github.com/neo4j-product-examples/snowflake-graph-analytics/tree/main/modeling-disruptions-subways

### What Is Neo4j Graph Analytics For Snowflake? 

> Neo4j helps organizations find hidden relationships and patterns across billions of data connections deeply, easily, and quickly. **Neo4j Graph Analytics for Snowflake** brings the power of graph directly to Snowflake, allowing users to run 65+ ready-to-use algorithms on their data, all without leaving Snowflake! 

Reviewer: One sentence per line

### Genre Recommendation with IMDB Data

> The IMDB dataset contains a rich graph of movies, actors, and directors with their relationships. This dataset originates from [Graph Transformer Networks](https://github.com/seongjunyun/Graph_Transformer_Networks) and has been processed to be compatible with the Snowflake application `Neo4j Graph Analytics`.

Reviewer: nit: One sentence per line

Using Graph Analytics for Snowflake, we can build powerful recommendation systems that leverage the complex relationships between movies, actors, and directors. This approach can be applied to various recommendation scenarios including content discovery, collaborative filtering, and personalized suggestions.

### Prerequisites

> - The Native App [Neo4j Graph Analytics](https://app.snowflake.com/marketplace/listing/GZTDZH40CN) for Snowflake

Reviewer: ```suggestion
- The native app [Neo4j Graph Analytics](https://app.snowflake.com/marketplace/listing/GZTDZH40CN) for Snowflake
```

### What You Will Need

- A [Snowflake account](https://signup.snowflake.com/?utm_cta=quickstarts) with appropriate access to databases and schemas.
- Neo4j Graph Analytics application installed from the Snowflake marketplace. Access the marketplace via the menu bar on the left hand side of your screen.

### What You Will Build

- Graph embeddings for movie recommendation systems
- A notebook that performs genre recommendation using GraphSAGE on IMDB data

> ### What You Will Learn

Reviewer: Two blank lines before a new heading, here and elsewhere

- **Features**: All nodes contain 1256-dimensional feature vectors via the "plot_keywords" property
- **Labels**: Some MOVIE nodes have a 'GENRE' value of 0, 1, or 2 for classification tasks

This dataset provides a rich foundation for understanding movie relationships and building recommendation systems.

## Loading The Data

Duration: 5

Let's name our database `IMDB`. We'll load the IMDB dataset into Snowflake using the provided scripts.

A script `genre-recommendation/upload_imdb_to_snowflake.py` uploads the data into a Snowflake account.

**Usage:**
```bash
uv sync

# Set --snowflake_account option in form of "ABCDEFG-MY_ACCOUNT"
uv run ./upload_imdb_to_snowflake.py --snowflake_account SNOWFLAKE_ACCOUNT \
                                     --snowflake_user SNOWFLAKE_USER \
                                     --snowflake_password SNOWFLAKE_PASSWORD \
                                     --snowflake_role SNOWFLAKE_ROLE \
                                     --snowflake_warehouse SNOWFLAKE_WAREHOUSE \
                                     --snowflake_database SNOWFLAKE_DATABASE \
                                     --snowflake_schema SNOWFLAKE_SCHEMA
```

> or if you have a connection set up in Snowflake. See [Snowflake documentation](https://docs.snowflake.com/en/developer-guide/snowpark/python/creating-session#connect-by-using-the-connections-toml-file).:

Reviewer: ```suggestion
If you have a connection config file for Snowflake you can select one of your configured connections. See [Snowflake documentation](https://docs.snowflake.com/en/developer-guide/snowpark/python/creating-session#connect-by-using-the-connections-toml-file).
```

```bash
uv run ./upload_imdb_to_snowflake.py --connection_name=my_connection
```

## Setting Up

### Import The Notebook

- We've provided a notebook to walk you through each SQL and Python step
- The notebook demonstrates GraphSAGE classification and embedding generation
- Don't forget to install required Python packages before you run

### Permissions

> Before we run our algorithms, we need to set the proper permissions. Ensure you are using `accountadmin` to grant and create roles:

Reviewer: On SPL

Author: on SPL? could you please elaborate?

Reviewer: One (S)entence (P)er (L)ine 😀 its my own acronym 🥇

GRANT SELECT ON ALL VIEWS IN SCHEMA genre_classification_db.imdb TO DATABASE ROLE graph_analytics_db_role;
GRANT SELECT ON FUTURE TABLES IN SCHEMA genre_classification_db.results TO DATABASE ROLE graph_analytics_db_role;
GRANT SELECT ON FUTURE TABLES IN SCHEMA genre_classification_db.imdb TO DATABASE ROLE graph_analytics_db_role;
GRANT SELECT ON FUTURE VIEWS IN SCHEMA genre_classification_db.imdb TO DATABASE ROLE graph_analytics_db_role;
GRANT CREATE TABLE ON SCHEMA genre_classification_db.results TO DATABASE ROLE graph_analytics_db_role;
GRANT CREATE TABLE ON SCHEMA genre_classification_db.imdb TO DATABASE ROLE graph_analytics_db_role;

-- Grant database role to application
GRANT DATABASE ROLE graph_analytics_db_role TO APPLICATION NEO4J_GRAPH_ANALYTICS;

-- Grant consumer role access to tables created by the application
GRANT USAGE ON DATABASE genre_classification_db TO ROLE gds_role;
GRANT USAGE ON SCHEMA genre_classification_db.results TO ROLE gds_role;
GRANT SELECT ON FUTURE TABLES IN SCHEMA genre_classification_db.results TO ROLE gds_role;
```

Then switch to the role we created:

```
USE ROLE gds_role;
```

> ## Generating graph embeddings with GraphSAGE

Reviewer: Should this be "node embedding"?
That's what is used below and it seems more accurate to me

Duration: 20

> GraphSAGE is a powerful algorithm for generating node embeddings by sampling and aggregating features from a node's local neighborhood. This approach allows us to capture the structural and feature information of nodes in a graph, making it ideal for tasks like node classification and recommendation systems.

Reviewer: ```suggestion
Duration: 20

GraphSAGE is a powerful algorithm for generating node embeddings by sampling and aggregating features from a node's local neighborhood. This approach allows us to capture the structural and feature information of nodes in a graph, making it ideal for tasks like node classification and recommendation systems.
```

```
-- Prediction stage of the GraphSAGE unsupervised algorithm - computing embeddings
CALL NEO4J_GRAPH_ANALYTICS.graph.gs_unsup_predict('GPU_NV_S', {
    'project': {
        'defaultTablePrefix': 'genre_classification_db.imdb',
        'nodeTables': ['actor', 'director', 'movie_plot'],
        'relationshipTables': {
            'acted_in': {'sourceTable': 'actor', 'targetTable': 'movie_plot', 'orientation': 'UNDIRECTED'},
            'directed_in': {'sourceTable': 'director', 'targetTable': 'movie_plot', 'orientation': 'UNDIRECTED'}
        }
    },
    'compute': {
        'modelname': 'unsup-imdb'
    },
    'write': [{
        'nodeLabel': 'movie_plot',
        'outputTable': 'genre_classification_db.results.movie_embeddings'
    }]
});
```

The resulting table `results.movie_embeddings` will contain the node IDs and their corresponding embeddings.

```
SELECT * FROM genre_classification_db.results.movie_embeddings LIMIT 5;
```

The model can be dropped when it is no longer needed:

```
CALL NEO4J_GRAPH_ANALYTICS.graph.drop_model('unsup-imdb');
```

## Node Classification with GraphSAGE

Duration: 20

> GraphSAGE can also be used for node classification tasks. In this example, we will classify movies into genres based on their features and relationships in the graph.

Reviewer: ```suggestion
Duration: 20

GraphSAGE can also be used for node classification tasks. In this example, we will classify movies into genres based on their features and relationships in the graph.
```

need the blank line or it gets rendered `Duration: 20 GraphSAGE can ...`
